pkg/integrations/v2/eventhandler/eventhandler.go

```diff
@@ -0,0 +1,454 @@
+// Agent's Logs subsystem (embedded promtail)
+package eventhandler
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"os/signal"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clientcmd"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/agent/pkg/integrations/v2"
+	"github.com/grafana/agent/pkg/logs"
+	"github.com/grafana/loki/clients/pkg/promtail/api"
+	"github.com/grafana/loki/pkg/logproto"
+	"github.com/prometheus/common/model"
+)
+
+const (
+	cacheFileMode = 0600
+)
+
+// EventHandler watches for Kubernetes Event objects and hands them off to
+// Agent's logs subsystem (embedded promtail).
+type EventHandler struct {
+	LokiClient    *logs.Instance
+	Log           log.Logger
+	CacheFile     *os.File
+	CachePath     string
+	LastEvent     *ShippedEvents
+	InitEvent     *ShippedEvents
+	EventInformer cache.SharedIndexInformer
+	ClusterLabel  string
+	SendTimeout   time.Duration
+	sync.Mutex
+}
+
+// ShippedEvents stores a timestamp and map of event ResourceVersions shipped for that timestamp.
+// Used to avoid double-shipping events upon restart.
+type ShippedEvents struct {
+	// shipped event's timestamp
+	Timestamp time.Time `json:"ts"`
+	// map of event RVs (resource versions) already "shipped" (handed off) for this timestamp.
+	// this is to handle the case of a timestamp having multiple events,
+	// which happens quite frequently.
+	RvMap map[string]struct{} `json:"resourceVersion"`
+}
+
+func newEventHandler(l log.Logger, globals integrations.Globals, c *Config) (integrations.Integration, error) {
+	var (
+		config *rest.Config
+		err    error
+	)
+
+	if c.InCluster {
+		config, err = rest.InClusterConfig()
+	} else {
+		config, err = clientcmd.BuildConfigFromFlags("", c.KubeconfigPath)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	// get an informer
+	factory := informers.NewSharedInformerFactory(clientset, time.Duration(c.InformerResync)*time.Second)
+	eventInformer := factory.Core().V1().Events().Informer()
+
+	eh := &EventHandler{
+		LokiClient:    globals.Logs.Instance(c.LogsInstance),
+		Log:           l,
+		CachePath:     c.CachePath,
+		EventInformer: eventInformer,
+		ClusterLabel:  c.ClusterName,
+		SendTimeout:   time.Duration(c.SendTimeout) * time.Second,
+	}
+	// set the resource handler fns
+	eh.initInformer(eventInformer)
+	return eh, nil
+}
+
+// Initialize informer by setting event handler fns
+func (eh *EventHandler) initInformer(eventsInformer cache.SharedIndexInformer) {
+	eventsInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    eh.addEvent,
+		UpdateFunc: eh.updateEvent,
+		DeleteFunc: eh.deleteEvent,
```

Comment:
What would cause a v1.Event to get updated / deleted? Does an update trigger if a resource is updated? Or is that a new event?

Reply:
Deletes occur when the events "expire" and get cleaned out of etcd so we just ignore these.

For updates, they occur in the following cases:
- [Resync](https://github.com/grafana/agent/pull/1310/files/064edc8e163843b4f43995843887bbc465ab19d9#r792144036)
- If a Pod is crash looping, for example, and the event occurs multiple times, new events aren't created each time; an existing event object is updated with a new RV and a "count" field is incremented. For the purposes of this integration, we just ship a new entry to Loki (the count field is shipped too)

If a resource is updated (for example you modify a Pod), that would be a new set of events emitted by the controller _I think_ — will double check.